### PR TITLE
HOMS-154 Add orders filter by custom fields

### PR DESCRIPTION
**Pull request by TimAle**

### CODE REVIEW

File: app/classes/custom_fields/data_types/datetime_range.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        coerce_value(value)
+      rescue ArgumentError
+        errors[attribute_name] << t('invalid_value',
+                                    attribute_name: attribute_name,
+                                    type: 'DateTimeFilter',
+                                    value: value)
+      end
+
+      def coerce_value(value)
+        if value.is_a?(Hash)
```
Comment (sugarfree1) [inline app/classes/custom_fields/data_types/datetime_range.rb]:
```
unless value.is_a?(Hash)
  raise NotImplementedError
end

...
```

File: app/filters/list_orders_filter.rb [hunk 2]
```diff
@@ -48,6 +48,29 @@ def filter_by_user_id(rel, value)
...
+      if data[:custom_fields]
+        fields_definition = OrderType.find(data[:order_type_id]).fields
+
+        string_field_names = fields_definition.select { |_k, v| %w(boolean number string).include? v[:type] }.keys
+        date_field_names   = fields_definition.select { |_k, v| v[:type] == 'datetime' }.keys
+
+        filter_string_fields = data[:custom_fields].slice(*string_field_names).compact
+        filter_date_fields   = data[:custom_fields].slice(*date_field_names).compact
+
+        filter_date_fields.reduce(relation.where('data @> ?', filter_string_fields.to_json)) do |rel, (name, value)|
```
Comment (sugarfree1) [inline app/filters/list_orders_filter.rb]: Put this scope ```data @>?``` to model.

File: app/filters/list_orders_filter.rb [hunk 3]
```diff
@@ -48,6 +48,29 @@ def filter_by_user_id(rel, value)
...
+          filter_by_custom_date_field(rel, name, value)
+        end
+      else
+        relation
+      end
+    end
+
+    def filter_by_custom_date_field(rel, name, value)
+      rel.where('(data->>:field)::timestamp between coalesce(:from, (data->>:field)::timestamp) and coalesce(:to, (data->>:field)::timestamp)',
```
Comment (sugarfree1) [inline app/filters/list_orders_filter.rb]: Same here.

**Comment by TimAle** [comment]
UPD

**Comment by TimAle** [comment]
UPD
checkbox margin fixed

**Comment by TimAle** [comment]
UPD
checkbox margin  and tests fixed

File: app/assets/stylesheets/framework_and_overrides.sass [hunk 4]
```diff
@@ -153,3 +153,29 @@ input
   a
     cursor: pointer
 
+#custom_field_filter_container
+  margin-left: -30px
+
+  .custom_field_filter
```
Comment (sugarfree1) [inline app/assets/stylesheets/framework_and_overrides.sass]: Class names usually have hyphens.

File: app/assets/stylesheets/framework_and_overrides.sass [hunk 5]
```diff
@@ -153,3 +153,29 @@ input
   a
     cursor: pointer
 
+#custom_field_filter_container
```
Comment (sugarfree1) [inline app/assets/stylesheets/framework_and_overrides.sass]: It's a common practice to put hyphens in ids.

File: app/assets/stylesheets/framework_and_overrides.sass [hunk 6]
```diff
@@ -153,3 +153,29 @@ input
...
+      overflow: hidden
+      text-overflow: ellipsis
+
+    .label-checkbox
+      margin-bottom: 0.3em
+
+    .checkbox-margin
+      margin: 10px 5px 19px
+
+    .custom_field_filter-close
```
Comment (sugarfree1) [inline app/assets/stylesheets/framework_and_overrides.sass]: Mixing hyphens and underscores.

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -15,6 +15,22 @@ def order_types_for_select
     end
   end
 
+  def custom_fields_for_filter(order_type, attributes)
+      type_attributes = OrderType.find(order_type).fields.with_indifferent_access
```
Comment (sugarfree1) [inline app/helpers/application_helper.rb]: Wrong alignment.

**Comment by TimAle** [comment]
UPD

**Pull request merged**